File: src/example_fgen_basic/error_v/creation.py
"""
Wrappers of `m_error_v_creation` [TODO think about naming and x-referencing]

At the moment, all written by hand.
We will auto-generate this in future.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from example_fgen_basic.error_v import ErrorV
from example_fgen_basic.pyfgen_runtime.exceptions import CompiledExtensionNotFoundError

try:
    from example_fgen_basic._lib import (  # type: ignore
        m_error_v_w,
    )
except (ModuleNotFoundError, ImportError) as exc:  # pragma: no cover
    raise CompiledExtensionNotFoundError("example_fgen_basic._lib.m_error_v_w") from exc
try:
    from example_fgen_basic._lib import m_error_v_creation_w
except (ModuleNotFoundError, ImportError) as exc:  # pragma: no cover
    raise CompiledExtensionNotFoundError(
        "example_fgen_basic._lib.m_error_v_creation_w"
    ) from exc

if TYPE_CHECKING:
    from example_fgen_basic.typing import NP_ARRAY_OF_INT


def create_error(inv: int) -> ErrorV:
    """
    Create an error

    Parameters
    ----------
    inv
        Input value

        If odd, the error code is
        [NO_ERROR_CODE][example_fgen_basic.error_v.error_v.NO_ERROR_CODE].
        If even, the error code is 1.
        If a negative number is supplied, the error code is 2.

    Returns
    -------
    :
        Created error
    """
    # Get the result, but receiving an instance index rather than the object itself
    instance_index: int = m_error_v_creation_w.create_error(inv)

    # Initialise the result from the received index
    res = ErrorV.from_instance_index(instance_index)

    return res


def create_errors(invs: NP_ARRAY_OF_INT) -> tuple[ErrorV, ...]:
    """
    Create a number of errors

    Parameters
    ----------
    invs
        Input values from which to create errors

        For each value in `invs`,
        if the value is even, an error is created,
        if the value is odd, an error with a no error code is created.

    Returns
    -------
    :
        Created errors
    """
    # Get the result, but receiving an instance index rather than the object itself
    instance_indexes: NP_ARRAY_OF_INT = m_error_v_creation_w.create_errors(invs)

    # Initialise the result from the received index
    res = tuple(ErrorV.from_instance_index(i) for i in instance_indexes)

    # Tell Fortran to finalise the object on the Fortran side
    # (all data has been copied to Python now)
    m_error_v_w.finalise_instances(instance_indexes)

    return res

File: src/example_fgen_basic/error_v/creation.f90
!> Error creation
!>
!> A very basic demo to get the idea.
!
! TODO: discuss - we should probably have some convention for module names.
! The hard part is avoiding them becoming too long...
module m_error_v_creation

    use m_error_v, only: ErrorV, NO_ERROR_CODE

    implicit none
    private

    public :: create_error, create_errors

contains

    function create_error(inv) result(err)
        !! Create an error
        !!
        !! If an odd number is supplied, the error code is no error (TODO: cross-ref).
        !! If an even number is supplied, the error code is 1.
        !! If a negative number is supplied, the error code is 2.

        integer, intent(in) :: inv
        !! Value to use to create the error

        type(ErrorV) :: err
        !! Created error

        if (inv < 0) then
            err = ErrorV(code=2, message="Negative number supplied")
            return
        end if

        if (mod(inv, 2) .eq. 0) then
            err = ErrorV(code=1, message="Even number supplied")
        else
            err = ErrorV(code=NO_ERROR_CODE)
        end if

    end function create_error

    function create_errors(invs, n) result(errs)
        !! Create a number of errors
        !!
        !! If an odd number is supplied, the error code is no error (TODO: cross-ref).
        !! If an even number is supplied, the error code is 1.
        !! If a negative number is supplied, the error code is 2.

        integer, dimension(n), intent(in) :: invs
        !! Values to use to create the error

        integer, intent(in) :: n
        !! Number of values to create

        type(ErrorV), dimension(n) :: errs
        !! Created errors

        integer :: i

        do i = 1, n

            errs(i) = create_error(invs(i))

        end do

    end function create_errors

end module m_error_v_creation
